modules/contact/src/constraints/ApplyFrictionConstraintLMMechanicalContact.C

```diff
@@ -0,0 +1,295 @@
+        Real integrated_velocity = _node_to_integrated_velocity[_current_node];
+
+        // NCP: requirement that either there is no slip **or** slip velocity and
+        // frictional force exerted **by** the secondary side are in the same direction
+        // We use relative tangential displacement as a proxy for slip velocity
+        Real a;
+        if (integrated_velocity * _u_secondary[_qp] < 0)
+          a = -std::abs(integrated_velocity * _dt);
+        else
+          a = std::abs(integrated_velocity * _dt);
+
+        // NCP part 2: require that the frictional force can never exceed the frictional
+        // coefficient times the normal force.
+        DualNumber<Real, Real> dual_u_secondary(_u_secondary[_qp]);
+        dual_u_secondary.derivatives() = 1;
+        auto b = _mu * _contact_pressure - std::abs(dual_u_secondary);
+
+        if (_ncp_type == "fb")
+          return (a + b - std::sqrt(a * a + b * b + _epsilon)).derivatives();
+        else
+          return std::min(a, b).derivatives();
+      }
+    }
+  }
+  _projection_successful = false;
+  return 0.0;
+}
+
+Real
+ApplyFrictionConstraintLMMechanicalContact::computeQpOffDiagJacobian(
+    Moose::ConstraintJacobianType /*type*/, unsigned int jvar)
+{
+  std::map<dof_id_type, PenetrationInfo *>::iterator found =
+      _penetration_locator._penetration_info.find(_current_node->id());
+  if (found != _penetration_locator._penetration_info.end())
+  {
+    PenetrationInfo * pinfo = found->second;
+    if (pinfo != NULL)
+    {
+      if (_contact_pressure < _epsilon || std::isnan(_node_to_integrated_velocity[_current_node]))
+        return 0.;
+
+      // Our local dual number is going to depend on only three degrees of freedom: the secondary
+      // nodal dofs for disp_x (index 0), disp_y (index 1), the contact pressure (index 2),
+      // and the tangential Lagrange multiplier. The latter of course exists only on the secondary
+      // side
+      // Use *existing* derivative info from the weighted tangential velocity
+      typedef DualNumber<Real, DNDerivativeSize<4>> LocalDN;
+
+      Real integrated_velocity = _node_to_integrated_velocity[_current_node];
+      LocalDN v_dot_tan(integrated_velocity);
+      //
+      //      (*_jacobian).
+      //      Moose::derivInsert(
+      //          v_dot_tan.derivatives(),
+      //          0,
+      //          (*_jacobian)(_current_node->dof_number(0, _var.number(), 0),
+      //          _connected_dof_indices[_j]));
+      //      Moose::derivInsert(
+      //          v_dot_tan.derivatives(),
+      //          1,
+      //          (*_jacobian)(_current_node->dof_number(0, _var.number(), 0),
+      //          _connected_dof_indices[_j]));
+      //      Moose::derivInsert(
+      //          v_dot_tan.derivatives(),
+      //          2,
+      //          (*_jacobian)(_current_node->dof_number(0, _var.number(), 0),
+      //          _connected_dof_indices[_j]));
```

Comment:
@lindsayad This commented out code was triggering an uninitialized read error in valgrind after the block had been in contact for a while and it was breaking off contact (even though these derivative additions do help convergence). Would you please review when you get the chance?